shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -0,0 +1,275 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.renderer;
+
+import android.annotation.TargetApi;
+import android.graphics.Bitmap;
+import android.graphics.SurfaceTexture;
+import android.os.Build;
+import android.os.Handler;
+import android.support.annotation.NonNull;
+import android.view.Surface;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicLong;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.view.TextureRegistry;
+
+/**
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ *
+ * {@code FlutterRenderer} works in tandem with a provided {@link RenderSurface} to create an
+ * interactive Flutter UI.
+ *
+ * {@code FlutterRenderer} manages textures for rendering, and forwards messages to native Flutter
+ * code via JNI. The corresponding {@link RenderSurface} is used as a delegate to carry out
+ * certain actions on behalf of this {@code FlutterRenderer} within an Android view hierarchy.
+ *
+ * {@link FlutterView} is an implementation of a {@link RenderSurface}.
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+public class FlutterRenderer implements TextureRegistry {
+
+  private final FlutterJNI flutterJNI;
+  private final long nativeObjectReference;
+  private final AtomicLong nextTextureId = new AtomicLong(0L);
+  private RenderSurface renderSurface;
+
+  public FlutterRenderer(@NonNull FlutterJNI flutterJNI, long nativeObjectReference) {
+    this.flutterJNI = flutterJNI;
+    this.nativeObjectReference = nativeObjectReference;
+  }
+
+  public void attachToRenderSurface(@NonNull RenderSurface renderSurface) {
+    // TODO(mattcarroll): determine desired behavior when attaching to an already attached renderer
+    if (this.renderSurface != null) {
+      detachFromRenderSurface();
+    }
+
+    this.renderSurface = renderSurface;
+    this.flutterJNI.setRenderSurface(renderSurface);
+  }
+
+  public void detachFromRenderSurface() {
+    // TODO(mattcarroll): determine desired behavior if we're asked to detach without first being attached
+    if (this.renderSurface != null) {
+      surfaceDestroyed();
+      this.renderSurface = null;
+      this.flutterJNI.setRenderSurface(null);
+    }
+  }
+
+  public void addOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedListener listener) {
+    flutterJNI.addOnFirstFrameRenderedListener(listener);
+  }
+
+  public void removeOnFirstFrameRenderedListener(@NonNull OnFirstFrameRenderedListener listener) {
+    flutterJNI.removeOnFirstFrameRenderedListener(listener);
+  }
+
+  //------ START TextureRegistry IMPLEMENTATION -----
+  // TODO(mattcarroll): this method probably shouldn't be public. It's part of an obscure relationship
```

Comment:
Not sure what do you mean by this shouldn't be public... ? don't we need a public way for plugins to create SurfaceTexture objects for textures that are registered by Flutter...

Reply:
Is this a capability that we want any developer to be able to invoke?  I think the reason I wrote this is because my initial investigation seemed to find that only platform views needed this and thus we could hide it within Flutter.  Do you know of use-cases that need this behavior outside of our own internal behavior?

Reply:
We definitely want to give plugins an ability to write to Textures and display them in Flutter.
You'd need this for any video/camera stream (we use this in the video and camera plugins).
Any app that is doing it's own GL rendering and want to embed it in Flutter will need this (and there are apps that does that).

Reply:
Makes sense.  I removed the comment.